src/ast_helpers.hpp

```diff
@@ -0,0 +1,348 @@
+  // EO equalsLiteral
+
+  // ##########################################################################
+  // Returns whether [name] is the name of a pseudo-element
+  // that can be written with pseudo-class syntax (CSS2 vs CSS3):
+  // `:before`, `:after`, `:first-line`, or `:first-letter`
+  // ##########################################################################
+  inline bool isFakePseudoElement(const std::string& name)
+  {
+    return equalsLiteral("after", name)
+      || equalsLiteral("before", name)
+      || equalsLiteral("first-line", name)
+      || equalsLiteral("first-letter", name);
+  }
+
+  // ##########################################################################
+  // Names of pseudo selectors that take selectors as arguments,
+  // and that are subselectors of their arguments.
+  // For example, `.foo` is a superselector of `:matches(.foo)`.
+  // ##########################################################################
+  inline bool isSubselectorPseudo(const std::string& norm)
+  {
+    return equalsLiteral("any", norm)
+      || equalsLiteral("matches", norm)
+      || equalsLiteral("nth-child", norm)
+      || equalsLiteral("nth-last-child", norm);
+  }
+  // EO isSubselectorPseudo
+
+  // ###########################################################################
+  // Pseudo-class selectors that take unadorned selectors as arguments.
+  // ###########################################################################
+  inline bool isSelectorPseudoClass(const std::string& test)
+  {
+    return equalsLiteral("not", test)
+      || equalsLiteral("matches", test)
+      || equalsLiteral("current", test)
+      || equalsLiteral("any", test)
+      || equalsLiteral("has", test)
+      || equalsLiteral("host", test)
+      || equalsLiteral("host-context", test);
+  }
+  // EO isSelectorPseudoClass
+
+  // ###########################################################################
+  // Pseudo-element selectors that take unadorned selectors as arguments.
+  // ###########################################################################
+  inline bool isSelectorPseudoElement(const std::string& test)
+  {
+    return equalsLiteral("slotted", test);
+  }
+  // EO isSelectorPseudoElement
+
+  // ###########################################################################
+  // Pseudo-element selectors that has binominals
+  // ###########################################################################
+  inline bool isSelectorPseudoBinominal(const std::string& test)
+  {
+    return equalsLiteral("nth-child", test)
+      || equalsLiteral("nth-last-child", test);
+  }
+  // isSelectorPseudoBinominal
+
+  // ###########################################################################
+  // Returns [name] without a vendor prefix.
+  // If [name] has no vendor prefix, it's returned as-is.
+  // ###########################################################################
+  inline std::string unvendor(const std::string& name)
```

Comment:
Consider moving these to `util_string.cpp` and adding unit tests.

Reply:
Done